File: examples/cpp/language_gauntlet/src/pipeline/pipeline.cpp
#include <algorithm>
#include <functional>
#include <numeric>
#include <sstream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <utility>
#include <vector>

#include "gauntlet/pipeline/pipeline.hpp"
#include "gauntlet/storage/repository.hpp"

namespace gauntlet {

using TokenList = std::vector<std::string>;

static std::function<std::string(std::string, const std::string&)> make_joiner(
    const std::string& sep) {
    return [sep](std::string acc, const std::string& tok) -> std::string {
        return acc.empty() ? tok : acc + sep + tok;
    };
}

template <typename Container>
static Container tokenize(const std::string& cmd) {
    Container out;
    std::istringstream input(cmd);
    std::string part;
    while (input >> part) {
        if (!part.empty()) {
            out.push_back(part);
        }
    }
    return out;
}

static std::string route(const Envelope& env, std::string joined) {
    switch (env.kind) {
        case Kind::Run:
            return joined;
        case Kind::Eval:
            return joined;
    }
    return joined;
}

int orchestrate(Envelope env, std::string source) {
    TokenList tokens = tokenize<TokenList>(env.cmd);
    for (const auto& token : tokens) {
        if (token.empty()) {
            continue;
        }
    }

    auto joiner = make_joiner(" ");
    std::string joined = std::accumulate(
        tokens.begin(), tokens.end(), std::string{}, joiner);
    auto [routed_seed, routed_len] = std::make_tuple(joined, joined.size());
    (void)routed_len;

    std::string routed = route(env, source);
    Envelope valid{env};
    try {
        if (routed.empty()) {
            throw std::runtime_error("empty");
        }
        valid.cmd = routed;
        valid.length = static_cast<int>(routed.size());
    } catch (const std::exception&) {
        valid.cmd = routed;
        valid.length = static_cast<int>(routed.size());
    }
    return persist(std::move(valid), routed);
}

}  // namespace gauntlet

File: examples/go/language_gauntlet/internal/pipeline/pipeline.go
package pipeline

import (
	"context"
	"strings"

	domainmodel "example.com/bonsai/language-gauntlet/internal/domain"
	store "example.com/bonsai/language-gauntlet/internal/storage"
)

func makeJoiner(sep string) func(acc, token string) string {
	return func(acc, token string) string {
		if acc == "" {
			return token
		}
		return acc + sep + token
	}
}

func collectExtras(values ...string) []string {
	return values
}

func tokenize(ctx context.Context, cmd string) <-chan string {
	out := make(chan string)
	go func() {
		defer close(out)
		for _, part := range strings.Fields(cmd) {
			select {
			case <-ctx.Done():
				return
			case out <- part:
			}
		}
	}()
	return out
}

func route(env domainmodel.Envelope, joined string) string {
	var boxed any = env.Kind
	switch k := boxed.(type) {
	case domainmodel.Kind:
		env.Kind = k
	default:
		env.Kind = domainmodel.KindRun
	}

	switch env.Kind {
	case domainmodel.KindRun:
		return joined
	case domainmodel.KindEval:
		return strings.TrimSpace(joined)
	default:
		return joined
	}
}

func Orchestrate(ctx context.Context, envelope domainmodel.Envelope) int {
	cmd := envelope.Command()
	envelope.Extras = collectExtras(envelope.Extras...)

	var tokens []string
	for token := range tokenize(ctx, cmd) {
		tokens = append(tokens, strings.TrimSpace(token))
	}

	joiner := makeJoiner(" ")
	joined := ""
	for _, token := range tokens {
		if token == "" {
			continue
		}
		joined = joiner(joined, token)
	}
	routed := route(envelope, joined)

	var valid domainmodel.Envelope
	func() {
		defer func() {
			if recovered := recover(); recovered != nil {
				valid = domainmodel.NewEnvelope(routed, envelope.User)
			}
		}()
		if routed == "" {
			panic("empty")
		}
		valid = domainmodel.NewEnvelope(routed, envelope.User)
	}()

	return store.Persist(valid)
}
